process/physics.py

```diff
@@ -379,389 +498,966 @@ def conhas(alphaj, alphap, bt, delta95, eps, kappa95, p0, rmu0):
+      Phys. Plasmas 1 July 1999; 6 (7): 2834–2839. https://doi.org/10.1063/1.873240
+    - O. Sauter, C. Angioni, Y. R. Lin-Liu; Erratum:
+      Neoclassical conductivity and bootstrap current formulas for general axisymmetric equilibria and arbitrary collisionality regime
+      [Phys. Plasmas 6, 2834 (1999)]. Phys. Plasmas 1 December 2002; 9 (12): 5140. https://doi.org/10.1063/1.1517052
     """
-    zz = zef[j - 1]
-    zft = tpf(j, triang, sqeps)
-    _nues = nues(j, rmajor, zef, mu, sqeps, tempe, ne)
-    zdf = (1.0 + (1.0 - 0.1 * zft) * np.sqrt(_nues)) + (0.5 * (1.0 - zft) * _nues) / zz
-    zft /= zdf  # $f^{31}_{teff}(\nu_{e*})$, Eq.14b
-    zft2 = zft**2
-    zft3 = zft2 * zft
-    dcsa = (
-        ((1.0 + 1.4 / (zz + 1.0)) * zft)
-        - (1.9 / (zz + 1.0) * zft2)
-        + ((0.3 * zft3 + 0.2 * zft3 * zft) / (zz + 1.0))
+    # Prevents first element being 0
+    charge_profile = zeff[radial_elements - 1]
+
+    # Calculate trapped particle fraction
+    f_trapped = trapped_particle_fraction(radial_elements, triang, sqeps)
+
+    # Calculated electron collisionality; nu_e*
+    electron_collisionality = electron_collisionality_sauter(
+        radial_elements, rmajor, zeff, inverse_q, sqeps, tempe, ne
+    )
+
+    # $f^{31}_{teff}(\nu_{e*})$, Eq.14b
+    f31_teff = f_trapped / (
+        (1.0 + (1.0 - 0.1 * f_trapped) * np.sqrt(electron_collisionality))
+        + (0.5 * (1.0 - f_trapped) * electron_collisionality) / charge_profile
+    )
+
+    l31_coefficient = (
+        ((1.0 + 1.4 / (charge_profile + 1.0)) * f31_teff)
+        - (1.9 / (charge_profile + 1.0) * f31_teff**2)
+        + ((0.3 * f31_teff**3 + 0.2 * f31_teff**4) / (charge_profile + 1.0))
     )
 
     # Corrections suggested by Fable, 15/05/2015
-    return dcsa * beta_poloidal_local_total(
-        j, nr, rmajor, bt, ne, ni, tempe, tempi, mu, rho
+    return l31_coefficient * beta_poloidal_total_sauter(
+        radial_elements,
+        number_of_elements,
+        rmajor,
+        bt,
+        ne,
+        ni,
+        tempe,
+        tempi,
+        inverse_q,
+        rho,
     )
 
 
 @nb.jit(nopython=True, cache=True)
-def hcsa(j, nr, rmajor, bt, triang, ne, ni, tempe, tempi, mu, rho, zef, sqeps):
-    """Grad(ln(Te)) coefficient in the Sauter bootstrap scaling
-    author: P J Knight, CCFE, Culham Science Centre
-    j  : input integer : radial element index in range 2 to nr
-    nr : input integer : maximum value of j
-    This function calculates the coefficient scaling grad(ln(Te))
-    in the Sauter bootstrap current scaling.
-    Code by Angioni, 29th May 2002.
-    <P>The code was supplied by Emiliano Fable, IPP Garching
-    (private communication).
-    O. Sauter, C. Angioni and Y. R. Lin-Liu,
-    Physics of Plasmas <B>6</B> (1999) 2834
-    O. Sauter, C. Angioni and Y. R. Lin-Liu, (ERRATA)
-    Physics of Plasmas <B>9</B> (2002) 5140
+def calculate_l31_32_coefficient(
```

Comment:
Should this be an internal function (ie start the name with _)?

Reply:
Changed